### PR TITLE
SI-7779 Account for class name compactification in reflection

### PR DESCRIPTION
**Pull request by retronym**
We have to assume that the classes we are reflecting on were
compiled with the default value for -Xmax-classfile-name (255).

With this assumption, we can apply the same name compactification
as done in the regular compiler.

The REPL is particularly prone to generating long class names
with the '$iw' prefixes, so this is an important fix for runtime
reflection.

Also adds support for getting the runtime class of `O.type` if
`O` is a module.

Review by @xeno-by

### CODE REVIEW

File: src/reflect/scala/reflect/internal/StdNames.scala [hunk 1]
```diff
@@ -44,6 +44,7 @@ trait StdNames {
     }
   }
 
+  private[reflect] def compactifyName(orig: String): String = compactify(orig)
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/StdNames.scala:47]: Do you think it'd be useful to expose this in the public API?

Comment (retronym) [inline src/reflect/scala/reflect/internal/StdNames.scala:47]: I'd rather not.

**Comment by xeno-by** [comment]
lgtm

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 2]
```diff
@@ -1241,6 +1254,7 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
       case TypeRef(_, ArrayClass, List(elemtpe)) => jArrayClass(typeToJavaClass(elemtpe))
       case TypeRef(_, sym: ClassSymbol, _) => classToJava(sym.asClass)
       case tpe @ TypeRef(_, sym: AliasTypeSymbol, _) => typeToJavaClass(tpe.dealias)
+      case SingleType(_, sym: ModuleSymbol) => classToJava(sym.moduleClass.asInstanceOf[ClassSymbol])
```
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: asClass

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 3]
```diff
@@ -1187,6 +1198,8 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
         noClass
     }
 
+    private val PackageAndClassPattern = """(.*\.)(.*)$""".r
```
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: how about making this a local variable?

Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: i wonder whether it's possible to write a pattern that'd also work for empty packages

Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: e.g. like this

```
>>> re.match("^(.*\.)?(.*)$", "Foo").groups()
(None, 'Foo')
>>> re.match("^(.*\.)?(.*)$", "bar.Foo").groups()
('bar.', 'Foo')
>>> re.match("^(.*\.)?(.*)$", "bar.baz.Foo").groups()
('bar.baz.', 'Foo')
```

Comment (retronym) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: It's a member to avoid recompiling the regex repeatedly.

Comment (retronym) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: Not sure if that style of regex plays nicely with pattern matching.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 2 above)
Comment (retronym) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: done.

**Comment by retronym** [comment]
I think it might be possible to improve on this to avoid the assumption that the reflection setting of -Xmax-classfile-name corresponds to the setting use to compile the class being reflected on. We can look at the length of candidate class names and the presence of the marker `$$$$` to figure that out.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 3 above)
Comment (som-snytt) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:1201]: Group not matching yields null.

```
scala> val r = """(?:(.*)\.)?(.*)""".r
r: scala.util.matching.Regex = (?:(.*)\.)?(.*)

scala> "a.b.c.d" match { case r(x,y) => s"$x/$y" }
res0: String = a.b.c/d

scala> "d" match { case r(x,y) => s"$x/$y" }
res1: String = null/d
```

**Comment by retronym** [comment]
BTW, I'd like to do that auto-detection of the length as a followup, so we can get this fix out for people using the default setting.

**Pull request merged**